File: 010 栈的基本实现/使用链表实现.c
#include <stdio.h>
#include <stdlib.h>

typedef struct Node {
    int value;
    struct Node *next;
} Node;

Node * createNode(int value) {
    Node *n = (Node *)malloc(sizeof(Node));
    n->value = value;
    n->next = NULL;
    return n;
}

void clean(Node *node) {
    node->next = NULL;
}

int getTop(Node *node) {
    if (node->next != NULL) {
        return node->next->value;
    }
    return -10000;
}

void push(Node *node, int e) {
    Node *n = createNode(e);
    n->next = node->next;
    node->next = n;
}

int pop(Node *node) {
    if (node->next != NULL) {
        int value = node->next->value;
        node->next = node->next->next;
        return value;
    }
    return -10000;
}

int isEmpty(Node *node) {
    return node->next == NULL;
}

int main() {
    Node *stack = createNode(0);
    for (int i = 0; i < 20; i++) {
        int v = rand() % 100;
        push(stack, v);
        printf("%d ", v);
    }
    printf("\n");
    
    printf("%d \n", getTop(stack));
    printf("%d \n", isEmpty(stack));
    
    printf("%d \n", pop(stack));
    printf("%d \n", getTop(stack));
    printf("%d \n", isEmpty(stack));
    
    clean(stack);
    printf("%d \n", getTop(stack));
    printf("%d \n", isEmpty(stack));
    
    pop(stack);
    printf("%d \n", getTop(stack));
    printf("%d \n", isEmpty(stack));
    return 0;
}

File: 010 栈的基本实现/使用链表实现.swift
import Foundation

class Node<E> {
    var value: E?
    var next: Node?
}

struct Stack<E> {
    let maxsize: Int
    var top: Node<E>?
    private(set) var length = 0
    
    init(_ maxsize: Int = 20) {
        self.maxsize = maxsize
    }
    
    mutating func clear() {
        length = 0
        top = nil
    }
    
    func isEmpty() -> Bool {
        return top == nil
    }
    
    func getTop() -> E? {
        return top?.value
    }
    
    mutating func push(_ e: E) -> Bool {
        if length >= maxsize {
            return false
        }
        let node = Node<E>()
        node.value = e
        node.next = top
        
        top = node
        length += 1
        return true
    }
    
    mutating func pop() -> E? {
        let v = top?.value
        top = top?.next
        
        if length > 0 {
            length -= 1
        }
        return v
    }
    
    func traverse() {
        var values = [E]()
        
        var n = top
        while n != nil {
            values.append(n!.value!)
            n = n?.next
        }
        for v in values.reversed() {
            print(v, terminator: " ")
        }
        print()
    }
}

var stack = Stack<Int>()

for _ in 0..<20 {
    let v = Int.random(in: 0..<100)
    print(v, terminator: " ")
    stack.push(v)
}
print()

stack.traverse()

print(stack.getTop())
print(stack.isEmpty())
print(stack.length)

print(stack.pop())
stack.traverse()
print(stack.getTop())
print(stack.isEmpty())
print(stack.length)


stack.clear()
stack.traverse()
print(stack.isEmpty())
print(stack.length)

print(stack.pop())
stack.traverse()
print(stack.getTop())
print(stack.isEmpty())
print(stack.length)

